### PR TITLE
Merge from Prof Hu's repo

### PR DESCRIPTION
**Pull request by mrezavandVirtonomy**
merging newest updates from Prof. Hu's repo

### CODE REVIEW

File: SPHINXsys/src/shared/materials/elastic_solid.cpp [hunk 1]
```diff
@@ -53,11 +53,6 @@ namespace SPH
 		LinearElasticSolid(Real rho0, Real youngs_modulus, Real poisson_ratio) : ElasticSolid(rho0)
 	{
 		material_type_ = "LinearElasticSolid";
-		assignElasticMaterialParameters(youngs_modulus, poisson_ratio);
-	}
-	//=================================================================================================//
-	void LinearElasticSolid::assignElasticMaterialParameters(Real youngs_modulus, Real poisson_ratio)
```
Comment (FabienPean-Virtonomy) [inline SPHINXsys/src/shared/materials/elastic_solid.cpp:59]: @mrezavandVirtonomy @BenceVirtonomy @JohnVirtonomy This code was removed from upstream but is used by John. This is not a big, but it is a good example to highlight future difficulties that we may face there. We should discuss how to handle this kind of scenario in the future. Right now the two options are: 1) replacing it in our codebase 2) reintroducing it upstream

File: SPHINXsys/src/shared/io_system/in_output.h [hunk 2]
```diff
@@ -310,6 +310,22 @@ namespace SPH
 		VtuStringData _vtuData;
 	};
 
+	/**
+	 * @class SurfaceOnlyBodyStatesRecordingToVtu
+	 * @brief  Write files for surface particles of bodies
+	 * the output file is VTK XML format can visualized by ParaView
+	 * the data type vtkUnstructedGrid
+	 */
+	class SurfaceOnlyBodyStatesRecordingToVtu : public BodyStatesRecording
```
Comment (FabienPean-Virtonomy) [inline SPHINXsys/src/shared/io_system/in_output.h]: This was added to upstream commit faa88cab07866f9ba5e370654ab3b515c46c6432 from our side
Removed from upstream during merge in PR #60 
Passed through during this merge, yielding duplicate symbols.

File: SPHINXsys/src/shared/bodies/base_body.cpp [hunk 3]
```diff
@@ -73,13 +73,13 @@ namespace SPH
 		newly_updated_ = false;
 	}
 	//=================================================================================================//
-	void SPHBody::writeParticlesToVtpFile(std::ostream &output_file)
+	void SPHBody::writeParticlesToVtpFile(std::ofstream &output_file)
```
Comment (FabienPean-Virtonomy) [inline SPHINXsys/src/shared/bodies/base_body.cpp:76]: Breaking changes (all the ostream->ofstream) from upstream

File: SPHINXsys/src/for_3D_build/particles/solid_particles_supplementary.cpp [hunk 4]
```diff
@@ -12,88 +12,24 @@ namespace SPH {
 			exit(1);
 	}
 	//=================================================================================================//
-	Matd ElasticSolidParticles::get_GreenLagrange_strain(size_t particle_i)
```
Comment (FabienPean-Virtonomy) [inline SPHINXsys/src/for_3D_build/particles/solid_particles_supplementary.cpp]: This file should be restored, it has been overwritten from upstream (which was our own previous version merged upstream last year)

File: SPHINXsys/src/for_3D_build/geometries/image_shape.cpp [hunk 5]
```diff
@@ -1,5 +1,3 @@
-#ifndef __EMSCRIPTEN__
```
Comment (FabienPean-Virtonomy) [inline SPHINXsys/src/for_3D_build/geometries/image_shape.cpp:1]: Not sure we want to lose these kind of things

**Comment by FabienPean-Virtonomy** [comment]
It feels like a lot was erased from our side in the merge, I will restart the whole merge process to have a global picture

**Comment by mrezavandVirtonomy** [comment]
We will discuss Fabien's comments afterwards.
The changes are merged in another PR.

**Pull request closed without merging**